### PR TITLE
Allow retroactive storage class assigment to PVCs

### PR DESCRIPTION
**Pull request by RomanBednar**
#### What type of PR is this?
/kind feature
/kind api-change

#### What this PR does / why we need it:
See linked KEP for details.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
PersistentVolumeClaim objects are no longer left with storage class set to `nil` forever, but will be updated retroactively once any StorageClass is set or created as default.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/pull/3337/commits/c6d69a727ebf56681a7d615c4f6234e4de86a5ec#diff-54da445e206aec7fe7962396edc49ec525ba1d3034aa800fd3a4e93d28a739e7
```
https://github.com/kubernetes/enhancements/issues/3333

### CODE REVIEW

**Review by jsafrane — COMMENTED** [review]
* Why is the PR marked as WIP?
* Please fix gofmt.

File: pkg/apis/core/validation/validation.go [hunk 1]
```diff
@@ -2216,13 +2219,18 @@ func ValidatePersistentVolumeClaimUpdate(newPvc, oldPvc *core.PersistentVolumeCl
...
+func validateStorageClassUpgrade(oldAnnotations, newAnnotations map[string]string, oldScName, newScName *string, opts PersistentVolumeClaimSpecValidationOptions) bool {
+	oldAnnotation, oldAnnotationExist := oldAnnotations[core.BetaStorageClassAnnotation]
 	newScInAnnotation, newAnnotationExist := newAnnotations[core.BetaStorageClassAnnotation]
 	return oldAnnotationExist /* condition 1 */ &&
 		oldScName == nil /* condition 2*/ &&
-		(newScName != nil && *newScName == oldSc) /* condition 3 */ &&
-		(!newAnnotationExist || newScInAnnotation == oldSc) /* condition 4 */
+		(newScName != nil && *newScName == oldAnnotation) /* condition 3 */ &&
+		(!newAnnotationExist || newScInAnnotation == oldAnnotation) /* condition 4 */ ||
+		(oldScName == nil && (!oldAnnotationExist || newScName == &oldAnnotation) && opts.EnableRetroactiveDefaultStorageClass) /* condition 5*/
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/apis/core/validation/validation.go]: Done.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 2]
```diff
@@ -918,6 +928,36 @@ func (ctrl *PersistentVolumeController) updateVolumePhaseWithEvent(volume *v1.Pe
 	return newVol, nil
 }
 
+// updateClaimStorageClass looks up all storage classes and attempts to assign a storage class to the given claim.
+// The storage class assigned is always the default one and if there is none or there are more defaults the claim
+// will not get updated.
+func (ctrl *PersistentVolumeController) updateClaimStorageClass(claim *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, error) {
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: Done.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 3]
```diff
@@ -918,6 +928,36 @@ func (ctrl *PersistentVolumeController) updateVolumePhaseWithEvent(volume *v1.Pe
...
+	// [Unit test set 6]
+	defaultClasses := storage.StorageClassList{}
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: Done.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 4]
```diff
@@ -918,6 +928,36 @@ func (ctrl *PersistentVolumeController) updateVolumePhaseWithEvent(volume *v1.Pe
...
+	if classes, err := ctrl.classLister.List(labels.Everything()); err != nil {
+		return nil, fmt.Errorf("error getting storage classes: %q", err)
+	} else {
+		for _, class := range classes {
+			if storageutil.IsDefaultAnnotationText(class.ObjectMeta) == "Yes" {
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: Done.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 5]
```diff
@@ -918,6 +928,36 @@ func (ctrl *PersistentVolumeController) updateVolumePhaseWithEvent(volume *v1.Pe
...
+				defaultClasses.Items = append(defaultClasses.Items, *class)
+			}
+		}
+		if len(defaultClasses.Items) > 1 {
+			return claim, fmt.Errorf("can not assign storage class to PersistentVolumeClaim[%s]: more than one default exists", claimToClaimKey(claim))
+		} else if len(defaultClasses.Items) == 0 {
+			return claim, fmt.Errorf("can not assign storage class to PersistentVolumeClaim[%s]: no default exists", claimToClaimKey(claim))
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: Done.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 6]
```diff
@@ -350,6 +352,14 @@ func (ctrl *PersistentVolumeController) syncUnboundClaim(ctx context.Context, cl
 			klog.V(4).Infof("synchronizing unbound PersistentVolumeClaim[%s]: no volume found", claimToClaimKey(claim))
 			// No PV could be found
 			// OBSERVATION: pvc is "Pending", will retry
+
+			if utilfeature.DefaultFeatureGate.Enabled(features.RetroactiveDefaultStorageClass) {
+				klog.V(4).Infof("FeatureGate[%s] is enabled, attempting to assign storage class to unbound PersistentVolumeClaim[%s]", features.RetroactiveDefaultStorageClass, claimToClaimKey(claim))
+				if claim, err = ctrl.updateClaimStorageClass(claim); err != nil {
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: Yes, I've missed that. Adding unit test and the missing condition.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 7]
```diff
@@ -918,6 +928,36 @@ func (ctrl *PersistentVolumeController) updateVolumePhaseWithEvent(volume *v1.Pe
 	return newVol, nil
 }
 
+// updateClaimStorageClass looks up all storage classes and attempts to assign a storage class to the given claim.
+// The storage class assigned is always the default one and if there is none or there are more defaults the claim
+// will not get updated.
+func (ctrl *PersistentVolumeController) updateClaimStorageClass(claim *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, error) {
+	// [Unit test set 6]
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: Done. Looks like the next free is `15`.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 8]
```diff
@@ -350,6 +352,14 @@ func (ctrl *PersistentVolumeController) syncUnboundClaim(ctx context.Context, cl
 			klog.V(4).Infof("synchronizing unbound PersistentVolumeClaim[%s]: no volume found", claimToClaimKey(claim))
 			// No PV could be found
 			// OBSERVATION: pvc is "Pending", will retry
+
+			if utilfeature.DefaultFeatureGate.Enabled(features.RetroactiveDefaultStorageClass) {
+				klog.V(4).Infof("FeatureGate[%s] is enabled, attempting to assign storage class to unbound PersistentVolumeClaim[%s]", features.RetroactiveDefaultStorageClass, claimToClaimKey(claim))
+				if claim, err = ctrl.updateClaimStorageClass(claim); err != nil {
+					klog.V(2).Infof("error updating PersistentVolumeClaim[%s]: %v", claimToClaimKey(claim), err)
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: Done.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 9]
```diff
@@ -350,6 +350,15 @@ func (ctrl *PersistentVolumeController) syncUnboundClaim(ctx context.Context, cl
 			klog.V(4).Infof("synchronizing unbound PersistentVolumeClaim[%s]: no volume found", claimToClaimKey(claim))
 			// No PV could be found
 			// OBSERVATION: pvc is "Pending", will retry
+
+			if utilfeature.DefaultFeatureGate.Enabled(features.RetroactiveDefaultStorageClass) {
+				klog.V(4).Infof("FeatureGate[%s] is enabled, attempting to assign storage class to unbound PersistentVolumeClaim[%s]", features.RetroactiveDefaultStorageClass, claimToClaimKey(claim))
+				if claim, err = ctrl.assignDefaultStorageClass(claim); err != nil {
+					// Intentionally ignoring errors - unsuccessful storage class assignment is expected if there is no default class or PVC already has a storage class
+					klog.V(2).Infof("can't update PersistentVolumeClaim[%s]: %v", claimToClaimKey(claim), err)
```
Comment (jsafrane) [inline pkg/controller/volume/persistentvolume/pv_controller.go:358]: I should have spotted it earlier, claim update error should not be ignored.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 10]
```diff
@@ -918,6 +927,32 @@ func (ctrl *PersistentVolumeController) updateVolumePhaseWithEvent(volume *v1.Pe
 	return newVol, nil
 }
 
+// assignDefaultStorageClass looks up all storage classes and attempts to assign a storage class to the given claim.
+// The storage class assigned is always the default one and if there is none or there are more defaults the claim
+// will not get updated.
+func (ctrl *PersistentVolumeController) assignDefaultStorageClass(claim *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, error) {
+	// [Unit test set 15]
+	if claim.Spec.StorageClassName != nil {
+		return claim, fmt.Errorf("can not assign storage class to PersistentVolumeClaim[%s]: pvc already has StorageClass[%s]", claimToClaimKey(claim), *claim.Spec.StorageClassName)
```
Comment (jsafrane) [inline pkg/controller/volume/persistentvolume/pv_controller.go:936]: don't return error here

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 9 above)
Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: Done, removing the comment and returning error.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 11]
```diff
@@ -918,6 +927,32 @@ func (ctrl *PersistentVolumeController) updateVolumePhaseWithEvent(volume *v1.Pe
...
+	} else if class == nil {
+		return claim, fmt.Errorf("can not assign storage class to PersistentVolumeClaim[%s]: default storage class not found", claimToClaimKey(claim))
+	}
+
+	klog.V(4).Infof("assigning StorageClass[%s] to PersistentVolumeClaim[%s]", class.Name, claimToClaimKey(claim))
+	claim.Spec.StorageClassName = &class.Name
+	newClaim, err := ctrl.kubeClient.CoreV1().PersistentVolumeClaims(claim.GetNamespace()).Update(context.TODO(), claim, metav1.UpdateOptions{})
+	if err != nil {
+		return nil, err
+	}
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller.go:953]: Done.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 12]
```diff
@@ -918,6 +927,32 @@ func (ctrl *PersistentVolumeController) updateVolumePhaseWithEvent(volume *v1.Pe
...
+	// [Unit test set 15]
+	if claim.Spec.StorageClassName != nil {
+		return claim, fmt.Errorf("can not assign storage class to PersistentVolumeClaim[%s]: pvc already has StorageClass[%s]", claimToClaimKey(claim), *claim.Spec.StorageClassName)
+	}
+
+	class, err := util.GetDefaultClass(ctrl.classLister)
+	if err != nil {
+		return claim, err
+	} else if class == nil {
+		return claim, fmt.Errorf("can not assign storage class to PersistentVolumeClaim[%s]: default storage class not found", claimToClaimKey(claim))
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: Done.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 13]
```diff
@@ -918,6 +927,32 @@ func (ctrl *PersistentVolumeController) updateVolumePhaseWithEvent(volume *v1.Pe
...
+// will not get updated.
+func (ctrl *PersistentVolumeController) assignDefaultStorageClass(claim *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, error) {
+	// [Unit test set 15]
+	if claim.Spec.StorageClassName != nil {
+		return claim, fmt.Errorf("can not assign storage class to PersistentVolumeClaim[%s]: pvc already has StorageClass[%s]", claimToClaimKey(claim), *claim.Spec.StorageClassName)
+	}
+
+	class, err := util.GetDefaultClass(ctrl.classLister)
+	if err != nil {
+		return claim, err
```
[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: Done.

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 10 above)
Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: Done.

**Comment by msau42** [comment]
API changes lgtm
/label api-review

File: pkg/apis/core/validation/validation.go [hunk 14]
```diff
@@ -2169,7 +2172,9 @@ func ValidatePersistentVolumeClaimUpdate(newPvc, oldPvc *core.PersistentVolumeCl
 	}
 
 	if validateStorageClassUpgrade(oldPvcClone.Annotations, newPvcClone.Annotations,
-		oldPvcClone.Spec.StorageClassName, newPvcClone.Spec.StorageClassName) {
+		oldPvcClone.Spec.StorageClassName, newPvcClone.Spec.StorageClassName) ||
+		validateStorageClassUpgradeFromNil(oldPvcClone.Annotations,
```
Comment (jsafrane) [inline pkg/apis/core/validation/validation.go:2176]: The error thrown below will complain about immutable `Annotations[v1.BetaStorageClassAnnotation]`, while Spec.StorageClassName should be immutable. It needs to distinct errors.

File: pkg/apis/core/validation/validation.go [hunk 15]
```diff
@@ -2169,7 +2172,9 @@ func ValidatePersistentVolumeClaimUpdate(newPvc, oldPvc *core.PersistentVolumeCl
 	}
 
 	if validateStorageClassUpgrade(oldPvcClone.Annotations, newPvcClone.Annotations,
```
Comment (jsafrane) [inline pkg/apis/core/validation/validation.go:2174]: Better rename this to validateStorageClassUpgradeFromAnnotation

File: test/integration/volume/persistent_volumes_test.go [hunk 16]
```diff
@@ -1036,6 +1040,164 @@ func TestPersistentVolumeMultiPVsDiffAccessModes(t *testing.T) {
...
+		t.Fatalf("Unexpected error getting pv: %v", err)
+	}
+	if pv.Spec.ClaimRef == nil {
+		t.Fatalf("PV %s with \"\" storage class should have been bound to PVC %s that also has \"\" storage class", pv2.Name, pvc2.Name)
+	}
+	if pv.Spec.ClaimRef.Name != pvc2.Name {
+		t.Fatalf("Bind mismatch! Expected %s but got %s", pvc2.Name, pv.Spec.ClaimRef.Name)
+	}
+
+	// 3. Test PV with storage class "" does not bind to PVC with nil while default SC exists.
```
Comment (jsafrane) [inline test/integration/volume/persistent_volumes_test.go:1170]: please add a note why they should not bind (different access mode)

File: test/integration/volume/persistent_volumes_test.go [hunk 17]
```diff
@@ -1036,6 +1040,164 @@ func TestPersistentVolumeMultiPVsDiffAccessModes(t *testing.T) {
...
+	// wait for pv to be created
+	waitForPersistentVolumePhase(testClient, pv3.Name, watchPV, v1.VolumePending)
+
+	pv, err = testClient.CoreV1().PersistentVolumes().Get(context.TODO(), "pv-3", metav1.GetOptions{})
+	if err != nil {
+		t.Fatalf("Unexpected error getting pv: %v", err)
+	}
+	if pv.Spec.ClaimRef != nil {
+		t.Fatalf("PV %s shouldn't be bound", pv3.Name)
+	}
```
Comment (jsafrane) [inline test/integration/volume/persistent_volumes_test.go:1193]: Move this check below after PVC is bound - at this time, a buggy PV controller might not have enough time to bind the bad PV.

Comment (RomanBednar) [reply test/integration/volume/persistent_volumes_test.go:1193]: Done.

File: pkg/apis/core/validation/validation.go (hunk 15 above)
Comment (RomanBednar) [reply pkg/apis/core/validation/validation.go:2174]: Done.

File: pkg/apis/core/validation/validation.go (hunk 14 above)
Comment (RomanBednar) [reply pkg/apis/core/validation/validation.go:2176]: Done. Moved it below to have it's own error. Not sure if immutable field error would be ok, it does not hint about it being mutable under certain conditions. Do we need a new error?

**Comment by jsafrane** [comment]
/test pull-kubernetes-e2e-kind
a readiness probe test failed

kubernetes-unit failures look genuine, the last update broke them.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 18]
```diff
@@ -918,6 +926,35 @@ func (ctrl *PersistentVolumeController) updateVolumePhaseWithEvent(volume *v1.Pe
 	return newVol, nil
 }
 
+// assignDefaultStorageClass looks up all storage classes and attempts to assign a storage class to the given claim.
+// The storage class assigned is always the default one and if there is none or there are more defaults the claim
+// will not get updated.
+func (ctrl *PersistentVolumeController) assignDefaultStorageClass(claim *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, error) {
+	// [Unit test set 15]
```
Comment (thockin) [inline pkg/controller/volume/persistentvolume/pv_controller.go:933]: Putting test stuff like this here is inevitably going to get broken or disconnected - can we not?

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 19]
```diff
@@ -918,6 +926,35 @@ func (ctrl *PersistentVolumeController) updateVolumePhaseWithEvent(volume *v1.Pe
...
+	if claim.Spec.StorageClassName != nil {
+		klog.V(4).Infof("can not assign storage class to PersistentVolumeClaim[%s]: pvc already has StorageClass[%s]", claimToClaimKey(claim), *claim.Spec.StorageClassName)
```
Comment (thockin) [inline pkg/controller/volume/persistentvolume/pv_controller.go:935]: This makes it feel like an error, but it's not, is it?  Any PVC that correctly got a default will trigger this?  Or maybe reword so it doesn't seem like a failure?

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 20]
```diff
@@ -918,6 +926,35 @@ func (ctrl *PersistentVolumeController) updateVolumePhaseWithEvent(volume *v1.Pe
...
+		return claim, nil
+	}
+
+	class, err := util.GetDefaultClass(ctrl.classLister)
+	if err != nil {
+		klog.V(4).Infof("failed to get default storage class: %v", err)
+		return claim, nil
```
Comment (thockin) [inline pkg/controller/volume/persistentvolume/pv_controller.go:942]: Why are you not returning the error?

Should this be `Errorf()` ?

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 21]
```diff
@@ -918,6 +926,35 @@ func (ctrl *PersistentVolumeController) updateVolumePhaseWithEvent(volume *v1.Pe
 	return newVol, nil
 }
 
+// assignDefaultStorageClass looks up all storage classes and attempts to assign a storage class to the given claim.
+// The storage class assigned is always the default one and if there is none or there are more defaults the claim
+// will not get updated.
```
Comment (thockin) [inline pkg/controller/volume/persistentvolume/pv_controller.go:931]: Can you say a little more about the fact that this will actually update the API (not the passed-in object) and then return the net-new object?  I did not expect that from this comment.

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 22]
```diff
@@ -918,6 +926,35 @@ func (ctrl *PersistentVolumeController) updateVolumePhaseWithEvent(volume *v1.Pe
...
+func (ctrl *PersistentVolumeController) assignDefaultStorageClass(claim *v1.PersistentVolumeClaim) (*v1.PersistentVolumeClaim, error) {
```
Comment (thockin) [inline pkg/controller/volume/persistentvolume/pv_controller.go:932]: This will get called when a PVC is updated or when the periodic sync fires.  How often is that?  Is it good enough?  Or should we also watch StorageClasses and trigger something whenever the effective default changes?  That's a lot more work, but if this will only trigger every O(days) it seems kind of bad..

File: pkg/controller/volume/persistentvolume/pv_controller_test.go [hunk 23]
```diff
@@ -776,3 +792,121 @@ func TestModifyDeletionFinalizers(t *testing.T) {
...
+		},
+		{
+			storageClasses: []*storagev1.StorageClass{
+				makeDefaultStorageClass(classGold, &modeImmediate),
+				makeStorageClass(classSilver, &modeImmediate),
+			},
+			tests: []controllerTest{
+				{
+					"15-5 - pvc storage class is assigned retroactively if there is a default",
+					novolumes,
```
Comment (thockin) [inline pkg/controller/volume/persistentvolume/pv_controller_test.go:880]: This super-anonymouos declaration pattern makes it REALLY hard for outsiders (me) to read the test cases - opportunity for improvment

File: pkg/apis/core/validation/validation.go [hunk 24]
```diff
@@ -2178,6 +2181,10 @@ func ValidatePersistentVolumeClaimUpdate(newPvc, oldPvc *core.PersistentVolumeCl
 		allErrs = append(allErrs, ValidateImmutableAnnotation(newPvc.ObjectMeta.Annotations[v1.BetaStorageClassAnnotation], oldPvc.ObjectMeta.Annotations[v1.BetaStorageClassAnnotation], v1.BetaStorageClassAnnotation, field.NewPath("metadata"))...)
 	}
 
+	if !validateStorageClassUpgradeFromNil(oldPvcClone.Annotations, oldPvcClone.Spec.StorageClassName, newPvcClone.Spec.StorageClassName, opts) {
+		allErrs = append(allErrs, ValidateImmutableField(newPvc.Spec.StorageClassName, oldPvc.Spec.StorageClassName, field.NewPath("metadata").Child("storageClassName"))...)
```
Comment (jpbetz) [inline pkg/apis/core/validation/validation.go:2185]: This results in a "field is immutable" error, but since the PR establishes conditions where it is mutable, should this be a `field.Invalid` with a bit more of an explanation?

File: test/integration/volume/persistent_volumes_test.go [hunk 25]
```diff
@@ -1106,6 +1266,27 @@ func waitForAnyPersistentVolumeClaimPhase(w watch.Interface, phase v1.Persistent
...
 
+func waitForPersistentVolumeClaimStorageClass(client *clientset.Clientset, claimName, namespace, scName string, w watch.Interface) {
+	// Check if the claim already has the storage class
+	claim, err := client.CoreV1().PersistentVolumeClaims(namespace).Get(context.TODO(), claimName, metav1.GetOptions{})
+	if err == nil && *claim.Spec.StorageClassName == scName {
+		return
+	}
+
+	// Wait for the storage class
+	for {
```
Comment (jpbetz) [inline test/integration/volume/persistent_volumes_test.go:1277]: Put a timeout on this? xref `apimachinery/watch.go:waitForEvent`

**Review by thockin — COMMENTED** [review]
The validation tests are complex enough that we might look for a better way to express the full matrix, but I won't block on that.

File: pkg/apis/core/validation/validation.go (hunk 24 above)
Comment (thockin) [reply pkg/apis/core/validation/validation.go:2185]: Good question.  For node PodCIDR we return `field.Forbidden` for "sometimes allowed but not like this".  That's not clearly right either.

Not sure which is best here, but Immutable seems wrong..

File: pkg/volume/util/storageclass.go [hunk 26]
```diff
@@ -0,0 +1,76 @@
...
+	"k8s.io/klog/v2"
+)
+
+const (
+	// isDefaultStorageClassAnnotation represents a StorageClass annotation that
+	// marks a class as the default StorageClass
+	IsDefaultStorageClassAnnotation = "storageclass.kubernetes.io/is-default-class"
+
+	// betaIsDefaultStorageClassAnnotation is the beta version of IsDefaultStorageClassAnnotation.
+	// TODO: remove Beta when no longer used
```
[the remark this answers is not in the document]
Comment (msau42) [reply pkg/volume/util/storageclass.go:35]: I believe we can never remove beta annotations....

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 22 above)
Comment (msau42) [reply pkg/controller/volume/persistentvolume/pv_controller.go:932]: Resync is every 15 seconds, and unfortunately there is behavior that is depending on frequent resyncs to happen rather than setting up the appropriate watches. I would love to get rid of the resync some day... https://github.com/kubernetes/kubernetes/issues/85661

File: pkg/volume/util/storageclass.go (hunk 26 above)
[the remark this answers is not in the document]
Comment (thockin) [reply pkg/volume/util/storageclass.go:35]: We have removed many before.  In this case, are we willing to read the annotation and set the `className` field?

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 22 above)
Comment (thockin) [reply pkg/controller/volume/persistentvolume/pv_controller.go:932]: Sad!  :)

Can we add a comment somewhere that spells out "if we ever change the resync to a larger period, we might need to change how we set the default class on existing unbound claims".

File: pkg/controller/volume/persistentvolume/pv_controller_test.go (hunk 23 above)
Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller_test.go]: Adding field names.

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 21 above)
Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: Done - changed the description to make it more clear.

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 20 above)
Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller.go:942]: `GetDefaultClass` returns errors either when it fails to list SCs or when there is more than one default. 

Anyway, it should be safe to ignore these errors - adding a comment and a link to PR where the function is improved.

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 19 above)
Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: I removed the message as it's not too useful anyway.

File: pkg/controller/volume/persistentvolume/pv_controller.go (hunk 18 above)
Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller.go]: Sure, removing.

File: test/integration/volume/persistent_volumes_test.go (hunk 25 above)
Comment (RomanBednar) [reply test/integration/volume/persistent_volumes_test.go]: Removed for now to pass feature freeze, integration tests will be added during this week as it needs further debugging.

File: pkg/apis/core/validation/validation.go (hunk 24 above)
Comment (RomanBednar) [reply pkg/apis/core/validation/validation.go]: I haven't found any simple way to return more specific error yet. Added a TODO note to fix it during beta.

**Comment by thockin** [comment]
We should consider whther this needs to go beta or just go straight to GA in 1.26.  I'm not sure what we would get out of being beta.  Honestly, even being alpha is minimal value here, but let's not change that now :)

/lgtm
/approve

File: test/integration/volume/persistent_volumes_test.go (hunk 25 above)
Comment (RomanBednar) [reply test/integration/volume/persistent_volumes_test.go]: @jpbetz Added [here](https://github.com/kubernetes/kubernetes/pull/111746/files#diff-6d10cbc48ef971686deb638dbc66a7f0498de8d9480f5de1275bbde421d379a5R1303)

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 27]
```diff
@@ -350,6 +350,14 @@ func (ctrl *PersistentVolumeController) syncUnboundClaim(ctx context.Context, cl
 			klog.V(4).Infof("synchronizing unbound PersistentVolumeClaim[%s]: no volume found", claimToClaimKey(claim))
 			// No PV could be found
 			// OBSERVATION: pvc is "Pending", will retry
+
+			if utilfeature.DefaultFeatureGate.Enabled(features.RetroactiveDefaultStorageClass) {
+				klog.V(4).Infof("FeatureGate[%s] is enabled, attempting to assign storage class to unbound PersistentVolumeClaim[%s]", features.RetroactiveDefaultStorageClass, claimToClaimKey(claim))
+				if claim, err = ctrl.assignDefaultStorageClass(claim); err != nil {
```
Comment (liggitt) [inline pkg/controller/volume/persistentvolume/pv_controller.go:356]: if this updated the storage class, don't we need to restart syncUnboundClaim so we recheck things like IsDelayBindingMode and findBestMatchForClaim?

Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller.go:356]: PR opened here: https://github.com/kubernetes/kubernetes/pull/113137

File: pkg/controller/volume/persistentvolume/pv_controller.go [hunk 28]
```diff
@@ -918,6 +926,35 @@ func (ctrl *PersistentVolumeController) updateVolumePhaseWithEvent(volume *v1.Pe
...
+		klog.V(4).Infof("failed to get default storage class: %v", err)
+		return claim, nil
+	} else if class == nil {
+		klog.V(4).Infof("can not assign storage class to PersistentVolumeClaim[%s]: default storage class not found", claimToClaimKey(claim))
+		return claim, nil
+	}
+
+	klog.V(4).Infof("assigning StorageClass[%s] to PersistentVolumeClaim[%s]", class.Name, claimToClaimKey(claim))
+	claim.Spec.StorageClassName = &class.Name
+	newClaim, err := ctrl.kubeClient.CoreV1().PersistentVolumeClaims(claim.GetNamespace()).Update(context.TODO(), claim, metav1.UpdateOptions{})
```
[the remark this answers is not in the document]
Comment (jsafrane) [reply pkg/controller/volume/persistentvolume/pv_controller.go:950]: IIRC we relaxed PVC validation as part of this feature, it can be updated from `nil` to something.

[the remark this answers is not in the document]
Comment (RomanBednar) [reply pkg/controller/volume/persistentvolume/pv_controller.go:950]: Exactly, the exact conditions for update from nil are set here: https://github.com/kubernetes/kubernetes/pull/111467/files#diff-c713e8919642d873fdf48fe8fb6d43e5cb2f53fd601066ff53580ea655948f0dR2238-R2251
